modelling/src/main/java/org/axonframework/modelling/configuration/DefaultStatefulCommandHandlingModule.java

```diff
@@ -0,0 +1,168 @@
+
+/**
+ * Basis implementation of the {@link StatefulCommandHandlingModule}.
+ *
+ * @author Allard Buijze
+ * @author Mateusz Nowak
+ * @author Mitchell Herrijgers
+ * @author Steven van Beelen
+ * @since 5.0.0
+ */
+class DefaultStatefulCommandHandlingModule
+        extends AbstractConfigurer<StatefulCommandHandlingModule>
+        implements StatefulCommandHandlingModule,
+        StatefulCommandHandlingModule.SetupPhase,
+        StatefulCommandHandlingModule.CommandHandlerPhase,
+        StatefulCommandHandlingModule.EntityPhase {
+
+    private final String moduleName;
+    private final String stateManagerName;
+    private final String statefulCommandHandlingComponentName;
+    private final Map<String, EntityBuilder<?, ?>> entityBuilders;
+    private final Map<QualifiedName, ComponentFactory<StatefulCommandHandler>> handlerFactories;
+    private final List<ComponentFactory<CommandHandlingComponent>> handlingComponentFactories;
+
+    DefaultStatefulCommandHandlingModule(@Nonnull String moduleName) {
+        Assert.nonEmpty(moduleName, "The module name cannot be null");
+        this.moduleName = moduleName;
+        this.stateManagerName = "StateManager[" + moduleName + "]";
+        this.statefulCommandHandlingComponentName = "StatefulCommandHandlingComponent[" + moduleName + "]";
+        this.entityBuilders = new HashMap<>();
+        this.handlerFactories = new HashMap<>();
+        this.handlingComponentFactories = new ArrayList<>();
+    }
+
+    @Override
+    public CommandHandlerPhase commandHandlers() {
+        return this;
+    }
+
+    @Override
+    public CommandHandlerPhase commandHandler(@Nonnull QualifiedName commandName,
+                                              @Nonnull ComponentFactory<StatefulCommandHandler> commandHandlerBuilder) {
+        handlerFactories.put(requireNonNull(commandName, "The command name cannot be null."),
+                             requireNonNull(commandHandlerBuilder, "The command handler builder cannot be null."));
+        return this;
+    }
+
+    @Override
+    public CommandHandlerPhase commandHandlingComponent(
+            @Nonnull ComponentFactory<CommandHandlingComponent> handlingComponentBuilder
+    ) {
+        handlingComponentFactories.add(
+                requireNonNull(handlingComponentBuilder, "The command handling component builder cannot be null.")
+        );
+        return this;
+    }
+
+    @Override
+    public EntityPhase entities() {
+        return this;
+    }
+
+    @Override
+    public <I, E> EntityPhase entity(@Nonnull EntityBuilder<I, E> entityBuilder) {
+        requireNonNull(entityBuilder, "The entity builder cannot be null.");
+        entityBuilders.put(entityBuilder.entityName(), entityBuilder);
+        return this;
+    }
+
+    @Override
+    public StatefulCommandHandlingModule build() {
+        registerRepositories();
+        registerComponent(StateManager.class, stateManagerName, this::stateManagerFactory);
```

Comment:
This line made me think about hierarchical state managers.
So, that a module can access the StateManager of it's parent, if there is one defined there. If the current and parent module both have the same entity, the current wins. What do you think?
Basically to define application-wide state, or module state that applies to its children